src/player.cpp

```diff
@@ -2415,14 +2415,95 @@ bool Player::addVIPInternal(uint32_t vipGuid)
 	return VIPList.insert(vipGuid).second;
 }
 
-bool Player::editVIP(uint32_t vipGuid, const std::string& description, uint32_t icon, bool notify)
+bool Player::addVIPGroupInternal(uint32_t vipGroupId)
+{
+	if (VIPGroups.size() >= getMaxVIPGroups()) {
+		return false;
+	}
+
+	return VIPGroups.insert(vipGroupId).second;
+}
+
+bool Player::editVIP(uint32_t vipGuid, const std::string& description, uint32_t icon, bool notify,
+                     const std::vector<uint16_t>& groupIds)
 {
 	auto it = VIPList.find(vipGuid);
 	if (it == VIPList.end()) {
 		return false; // player is not in VIP
 	}
 
-	IOLoginData::editVIPEntry(accountNumber, vipGuid, description, icon, notify);
+	IOLoginData::editVIPEntry(accountNumber, vipGuid, description, icon, notify, groupIds);
+	return true;
+}
+
+bool Player::addVIPGroup(const std::string& name)
+{
+	if (VIPGroups.size() >= getMaxVIPGroups()) {
+		sendTextMessage(MESSAGE_STATUS_SMALL,
+		                "You have already reached the maximum of groups you can create yourself.");
+		return false;
+	}
+
+	const std::vector<std::string>& reservedNames = {"Friends", "Enemies", "Trading Partners"};
+	if (std::find(reservedNames.begin(), reservedNames.end(), name) != reservedNames.end()) {
+		sendTextMessage(MESSAGE_STATUS_SMALL, "You have selected an invalid group name. Please choose another one.");
+		return false;
+	}
+
+	if (IOLoginData::checkVIPGroupName(accountNumber, name)) {
+		sendTextMessage(MESSAGE_STATUS_SMALL, "A group with this name already exists. Please choose another name.");
+		return false;
+	}
+
+	uint32_t newId = IOLoginData::addVIPGroup(accountNumber, name, true);
+	auto result = VIPGroups.insert(newId);
+	if (!result.second) {
+		sendTextMessage(MESSAGE_STATUS_SMALL, "This group is already in your group list.");
+		return false;
+	}
+
+	if (client) {
+		client->sendVIPGroups();
+	}
+
+	return true;
+}
+
+bool Player::editVIPGroup(uint16_t vipGroupId, const std::string& name)
+{
+	const std::vector<std::string>& reservedNames = {"Friends", "Enemies", "Trading Partners"};
```

Comment:
```suggestion
	auto reservedNames = std::array{"Friends", "Enemies", "Trading Partners"};
```

It appears more than once, you may consider moving it to a constexpr outside these functions instead